src/value/csv.rs

```diff
@@ -0,0 +1,80 @@
+use csv;
+use error;
+use std::io;
+use value;
+use ordered_float;
+
+pub struct CsvSource<R>(csv::StringRecordsIntoIter<R>) where R: io::Read;
+
+pub struct CsvSink<W>(csv::Writer<W>) where W: io::Write;
+
+#[inline]
+pub fn source<R>(r: R) -> CsvSource<R> where R: io::Read
+{
+    CsvSource(csv::ReaderBuilder::new()
+                  .has_headers(false)
+                  .from_reader(r)
+                  .into_records())
+}
+
+#[inline]
+pub fn sink<W>(w: W) -> CsvSink<W> where W: io::Write
+{
+    CsvSink(csv::Writer::from_writer(w))
+}
+
+impl<R> value::Source for CsvSource<R> where R: io::Read
+{
+    #[inline]
+    fn read(&mut self) -> error::Result<Option<value::Value>> {
+        match self.0.next() {
+            Some(Ok(v)) => Ok(Some(value::Value::Sequence(
+                v.iter().map(|s| value::Value::String(s.to_string())).collect()
+            ))),
+            Some(Err(e)) => Err(error::Error::from(e)),
+            None => Ok(None),
+        }
+    }
+}
+
+impl<W> value::Sink for CsvSink<W> where W: io::Write
+{
+    #[inline]
+    fn write(&mut self, value: value::Value) -> error::Result<()> {
+        match value {
+            value::Value::Sequence(seq) => {
+                let record: Vec<String> = seq.into_iter().map(value_to_csv).collect();
+                self.0.write_record(record)?;
+                Ok(())
+            }
+            x => bail!("csv can only output sequences, got: {:?}", x)
+        }
+    }
+}
+
+fn value_to_csv(value: value::Value) -> String {
+    match value {
+        value::Value::Unit => panic!("csv cannot output nested Unit"),
```

Comment:
You can make this function return `error::Result<String>` instead of `String` which lets you use `bail!()` (which internally expands to `return Err(...)`)

Reply:
Yes, but the function is used as a map on a vector, which makes this messy.

Reply:
you can say `collect::<Result<Vec<_>>()?` to get a Vec as the result.